File: gp_system_collections.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.AspNetCore.Mvc;

namespace API
{
    // Enumler
    public enum DatabaseProvider { MongoDB, Postgres, Firebase, Memory }
    public enum DatabaseEnvironment { PRODUCTION, STAGING, TEST, DEVELOPMENT }
    public enum ActionType { Read, Create, Update, Delete }

    // Modeller
    public class GlobalPolicyRule
    {
        public ActionType Action { get; set; }
        public string Condition { get; set; } 
        public bool Allow { get; set; }
    }

    public class IndexDefinition
    {
        public Dictionary<string, int> Keys { get; set; } = new Dictionary<string, int>();
        public bool? Unique { get; set; }
        public bool? Sparse { get; set; }
        public int? ExpireAfterSeconds { get; set; }
    }

    public class CollectionRule
    {
        public string ListRule { get; set; }
        public string ViewRule { get; set; }
        public string CreateRule { get; set; }
        public string UpdateRule { get; set; }
        public string DeleteRule { get; set; }
    }

    public class SubCollection
    {
        public string Name { get; set; }
        public string Description { get; set; }
        public string Parent { get; set; }
    }

    public class Collection
    {
        public string Name { get; set; }
        public string Description { get; set; }
        public bool? System { get; set; }
        public long? CreatedAt { get; set; }
        public List<SubCollection> SubCollections { get; set; } = new List<SubCollection>();
        public CollectionRule Rules { get; set; }
        public List<IndexDefinition> Indexes { get; set; } = new List<IndexDefinition>();
    }

    // Servis (in-memory)
    public class CollectionService
    {
        private readonly List<Collection> _collections = new List<Collection>();

        public IEnumerable<Collection> GetAll() => _collections;

        public Collection GetByName(string name) =>
            _collections.FirstOrDefault(c => c.Name == name);

        public Collection Create(Collection collection)
        {
            if (_collections.Any(c => c.Name == collection.Name))
                throw new InvalidOperationException($"Collection '{collection.Name}' already exists.");

            collection.CreatedAt = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
            _collections.Add(collection);
            return collection;
        }

        public Collection Update(string name, Collection updated)
        {
            var collection = GetByName(name);
            if (collection == null)
                throw new KeyNotFoundException($"Collection '{name}' not found.");

            collection.Description = updated.Description;
            collection.System = updated.System;
            collection.SubCollections = updated.SubCollections;
            collection.Rules = updated.Rules;
            collection.Indexes = updated.Indexes;

            return collection;
        }

        public void Delete(string name)
        {
            var collection = GetByName(name);
            if (collection == null)
                throw new KeyNotFoundException($"Collection '{name}' not found.");

            _collections.Remove(collection);
        }
    }

    // Controller
    [ApiController]
    [Route("api/[controller]")]
    public class CollectionsController : ControllerBase
    {
        private readonly CollectionService _service = new CollectionService();

        [HttpGet]
        public ActionResult<IEnumerable<Collection>> GetAll()
        {
            return Ok(_service.GetAll());
        }

        [HttpGet("{name}")]
        public ActionResult<Collection> Get(string name)
        {
            var collection = _service.GetByName(name);
            if (collection == null) return NotFound();
            return Ok(collection);
        }

        [HttpPost]
        public ActionResult<Collection> Create([FromBody] Collection collection)
        {
            try
            {
                var created = _service.Create(collection);
                return CreatedAtAction(nameof(Get), new { name = created.Name }, created);
            }
            catch (InvalidOperationException ex)
            {
                return Conflict(ex.Message);
            }
        }

        [HttpPut("{name}")]
        public ActionResult<Collection> Update(string name, [FromBody] Collection updated)
        {
            try
            {
                var result = _service.Update(name, updated);
                return Ok(result);
            }
            catch (KeyNotFoundException)
            {
                return NotFound();
            }
        }

        [HttpDelete("{name}")]
        public ActionResult Delete(string name)
        {
            try
            {
                _service.Delete(name);
                return NoContent();
            }
            catch (KeyNotFoundException)
            {
                return NotFound();
            }
        }
    }
}

File: gp_system_collections.ts
const SYSTEM_COLLECTION = "gp_system_collections";

export type DatabaseProvider = "mongodb" | "postgres" | "firebase" | "memory";

export interface GlobalPolicyRule {
  action: "read" | "create" | "update" | "delete";
  condition: string; // "@request.auth.id != null" gibi
  allow: boolean;
}

export type DatabaseEnvironment =
  | "PRODUCTION"
  | "STAGING"
  | "TEST"
  | "DEVELOPMENT";

export interface IndexDefinition {
  keys: Record<string, 1 | -1>; // 1: ascending, -1: descending
  unique?: boolean;
  sparse?: boolean;
  expireAfterSeconds?: number; // TTL
}

export interface CollectionRule {
  listRule?: string;
  viewRule?: string;
  createRule?: string;
  updateRule?: string;
  deleteRule?: string;
}

export interface SubCollection {
  name: string;
  description?: string;
  parent?: string;
}

export interface Collection {
  name: string;
  description?: string;
  system?: boolean; // Sistem koleksiyonu mu?
  createdAt?: number; // Timestamp (ms)
  subCollections?: SubCollection[];
  rules?: CollectionRule;
  indexes?: IndexDefinition[];
}

export interface Database {
  name: string;
  collections: Collection[];
  provider: DatabaseProvider;
  environment: DatabaseEnvironment;
  globalPolicy?: GlobalPolicyRule[];
}

export const collections: Collection[] = [
  {
    name: "car_listings",
    description: "Araç ilanları",
    subCollections: [
      {
        name: "car_listings",
        parent: "car_listings",
        description: "Araç fotoğrafları",
      },
      {
        name: "car_listings_details",
        parent: "car_listings",
        description: "Araç İlan Detayları",
      },
      {
        name: "car_listings_photo_datas",
        parent: "car_listings",
        description: "Araç fotoğraf veri detayları",
      },
      {
        name: "car_listings_metrics",
        parent: "car_listings",
        description: "Araç ilan metrikleri",
      },
      {
        name: "car_listings_notes",
        parent: "car_listings",
        description: "İlan notları",
      },
      {
        name: "car_listings_equipment",
        parent: "car_listings",
        description: "Seçilen donanım özellikleri",
      },
      {
        name: "car_listings_seller_info",
        parent: "car_listings",
        description: "İlan satıcı bilgileri",
      },
      {
        name: "car_listings_messaging",
        parent: "car_listings",
        description: "İlan mesajlaşma geçmişi",
      },
    ],
    rules: {
      createRule: "@request.auth.id != ''",
      updateRule: "@request.auth.id = @record.author",
      deleteRule: "@request.auth.id = @record.author",
    },
    indexes: [
      { keys: { listingId: 1 }, unique: true },
      { keys: { createdAt: -1 } },
    ],
  },
];

export const databases: Database[] = [
  {
    name: "db",
    provider: "mongodb",
    environment: "PRODUCTION",
    globalPolicy: [
      { action: "read", condition: "@request.auth.id != null", allow: true },
      {
        action: "create",
        condition: "@request.auth.roles contains 'editor'",
        allow: true,
      },
    ],

    collections: [
      {
        name: "car_listings",
        description: "Araç ilanları",
        system: false,
        createdAt: Date.now(),
        subCollections: [
          { name: "car_listings_photos", parent: "car_listings" },
          { name: "car_listings_videos", parent: "car_listings" },
        ],
        rules: {
          listRule: "@request.auth.id != null",
          createRule: "@request.auth.id != ''",
        },
        indexes: [
          { keys: { listingId: 1 }, unique: true },
          { keys: { createdAt: -1 } },
        ],
      },
      {
        name: "medias",
        description: "Medya dosyaları",
        system: true,
        createdAt: Date.now(),
        subCollections: [{ name: "media_thumbnails", parent: "medias" }],
      },
    ],
  },
  {
    name: "staging-db",
    provider: "mongodb",
    environment: "PRODUCTION",
    collections: [
      {
        name: "car_listings",
        description: "Araç ilanları",
        system: false,
        createdAt: Date.now(),
        subCollections: [
          { name: "car_listings_photos", parent: "car_listings" },
          { name: "car_listings_videos", parent: "car_listings" },
        ],
        rules: {
          listRule: "@request.auth.id != null",
          createRule: "@request.auth.id != ''",
        },
        indexes: [
          { keys: { listingId: 1 }, unique: true },
          { keys: { createdAt: -1 } },
        ],
      },
      {
        name: "medias",
        description: "Medya dosyaları",
        system: true,
        createdAt: Date.now(),
        subCollections: [{ name: "media_thumbnails", parent: "medias" }],
      },
    ],
  },
];

// helper.ts

export const getSubCollectionNames = (collectionName: string): string[] => {
  const col = collections.find((c) => c.name === collectionName);
  if (!col || !col.subCollections) return [];
  return col.subCollections.map((sub) => sub.name);
};

// Kullanım örneği
const mediasSubs = getSubCollectionNames("medias");
console.log(mediasSubs);
// → ["media_details", "media_thumbnails"]

const carListingsSubs = getSubCollectionNames("car_listings");
console.log(carListingsSubs);
// → ["photos", "videos"]
